cupy/_core/_gufuncs.py

```diff
@@ -405,120 +420,167 @@ def _apply_func_to_inputs(self, func, dim, sizes, dims, args, outs):
+                n_outs = [o[i] for o in outs]
+                self._apply_func_to_inputs(
+                    func, outer_shape[1:], n_args, n_outs)
+
+    def _update_dims(self, i, core_dims, cd, length):
+        # Helper to update dimensions, just to not repeat error messages.
+        name = cd[0]
+        prev_length = core_dims.get(name, None)
+        if prev_length is None:
+            if cd[2] and length == 1:
+                return  # broadcastable, other operand may set it.
+            core_dims[name] = length
+        elif length == 1 and cd[2]:  # cd[2] indicates broadcastable
+            # broadcastable core-dim of size 1, ignore it.
+            return
+        elif prev_length != length:
+            if prev_length == -1 or length == -1:
                 raise ValueError(
-                    f'Dimension {dim} with different lengths in arrays'
-                )
+                    'An optional core-dimension must be skipped in '
+                    'all or no inputs.')
+            raise ValueError(
+                f'Input operand {i} has mismatch in its core '
+                f'dimension {name} with signature {self._signature} '
+                f'({prev_length} != {length}).')
+
+    def _setup_operands(
+            self, args, in_axes, outs, out_axes, keepdims,
+            ret_dtype, filter_order, casting):
+        """Set up the operands for the function call, this needs to figure out
+        the actual core axes and shapes and make sure core-dimensions match.
+        We then transpose the core dimensions to the end for the actual
+        operation.
+
+        This function also sets up the output operands, note that there are two
+        versions. The untransposed ones for user-return and the transposed ones
+        for the actual result.
+        """
+        core_dims = {}
+        outer_shapes = []
+        transposals = []
+        for i, (arg, axs, coredims) in enumerate(
+                zip(args, in_axes, self._input_coredimss)):
+            ndim = arg.ndim
+            n_core_dims = max(self._min_dims[i], min(ndim, len(coredims)))
+            transpose = self._get_transpose(i, axs, ndim, n_core_dims)
+            if transpose is None:
+                outer_shape = arg.shape[:-n_core_dims]
+                core_shape = arg.shape[-n_core_dims:]
+            else:
+                outer_shape = tuple(arg.shape[i]
+                                    for i in transpose[:-n_core_dims])
+                core_shape = tuple(arg.shape[i]
+                                   for i in transpose[-n_core_dims:])
+
+            n_skipped = len(coredims) - n_core_dims
+            ommitted_coredims = []
+            i_coredim = 0
+            for length in core_shape:
+                # Process the core dimension shape and store the result into
+                # `core_dims` for each dimension.
+
+                # If there are optional coredims skip them from the front
```

Comment:
The old code supported at-front or at-end, this supports both.  The way this works in NumPy is that we skip at the front, which is done here.

(Could probably support mixing `n?` and `n` properly, I don't think it is but doens't matter.)